tools/src/bin/build-blame.rs

```diff
@@ -0,0 +1,357 @@
+    blame_repo: &git2::Repository,
+    blame_parents: &[git2::Commit],
+    mut path: PathBuf,
+) -> Result<(), git2::Error> {
+    let tree_at_path = if path == PathBuf::new() {
+        commit.tree()?
+    } else {
+        commit.tree()?.get_path(&path)?.to_object(git_repo)?.peel_to_tree()?
+    };
+    'outer: for entry in tree_at_path.iter() {
+        path.push(entry.name().unwrap());
+        for (i, parent) in commit.parents().enumerate() {
+            if let Ok(parent_entry) = parent.tree()?.get_path(&path) {
+                if parent_entry.id() == entry.id() {
+                    // Item at `path` is the same in the tree for `commit` as in the tree
+                    // for `parent`, so the blame must be the same too
+                    let blame_parent_entry = blame_parents[i].tree()?.get_path(&path)?;
+                    builder.insert(entry.name().unwrap(), blame_parent_entry.id(), entry.filemode())?;
+                    path.pop();
+                    continue 'outer;
+                }
+            }
+        }
+
+        match entry.kind() {
+            Some(ObjectType::Blob) => {
+                let blame_text = blame_for_path(
+                    file_movement,
+                    git_repo,
+                    commit,
+                    blame_repo,
+                    blame_parents,
+                    &path
+                )?;
+                builder.insert(entry.name().unwrap(), blame_repo.blob(&blame_text.as_bytes())?, entry.filemode())?;
+            }
+            Some(ObjectType::Commit) => {
+                // This is a submodule, just treat it as an empty dir. We could
+                // probably also skip over it entirely.
+                let entry_builder = blame_repo.treebuilder(None)?;
+                builder.insert(entry.name().unwrap(), entry_builder.write()?, entry.filemode())?;
+            }
+            Some(ObjectType::Tree) => {
+                let mut entry_builder = blame_repo.treebuilder(None)?;
+                build_blame_tree(
+                    &mut entry_builder,
+                    file_movement,
+                    git_repo,
+                    commit,
+                    blame_repo,
+                    blame_parents,
+                    path.clone(),
+                )?;
+                builder.insert(entry.name().unwrap(), entry_builder.write()?, entry.filemode())?;
+            }
+            _ => {
+                panic!("Unexpected entry kind {:?} found in tree for commit {:?} at path {:?}", entry.kind(), commit.id(), path);
+            }
+        };
+
+        path.pop();
+    }
+
+    Ok(())
+}
+
+fn main() {
+    env_logger::init();
+
+    let args: Vec<_> = env::args().collect();
+    let git_repo = Repository::open(&args[1]).unwrap();
+    let blame_repo = Repository::open(&args[2]).unwrap();
+    let use_cinnabar = env::var("CINNABAR").map(|v| v == "1").unwrap_or(false);
+    let blame_ref = env::var("BLAME_REF").ok().unwrap_or("HEAD".to_string());
```

Comment:
nit: `unwrap_or_else`? Though arguably not in hot code enough to matter at all.

Reply:
Yeah this just runs once. Left it as-is.